File: sys/src/nonlinear/levenberg_marquardt_params.cc
#include "levenberg_marquardt_params.h"

namespace gtsam {

std::unique_ptr<LevenbergMarquardtParams> default_levenberg_marquardt_params() {
  return std::make_unique<LevenbergMarquardtParams>();
}

void levenberg_marquardt_params_set_max_iterations(
    LevenbergMarquardtParams &params, uint32_t value) {
  return params.setMaxIterations(value);
}

void levenberg_marquardt_params_set_lambda_upper_bound(
    LevenbergMarquardtParams &params, double value)
{
    return params.setlambdaUpperBound(value);
}

void levenberg_marquardt_params_set_lambda_lower_bound(
    LevenbergMarquardtParams &params, double value)
{
    return params.setlambdaLowerBound(value);
}

void levenberg_marquardt_params_set_diagonal_damping(
    LevenbergMarquardtParams &params, bool flag)
{
    return params.setDiagonalDamping(flag);
}

void levenberg_marquardt_params_set_relative_error_to_l(
    LevenbergMarquardtParams &params, double value)
{
    return params.setRelativeErrorTol(value);
}

void levenberg_marquardt_params_set_absolute_error_to_l(
    LevenbergMarquardtParams &params, double value)
{
    return params.setAbsoluteErrorTol(value);
}

} // namespace gtsam

File: src/nonlinear/levenberg_marquardt_params.rs
use cxx::UniquePtr;

pub struct LevenbergMarquardtParams {
    pub(super) inner: UniquePtr<::sys::LevenbergMarquardtParams>,
}

impl Default for LevenbergMarquardtParams {
    fn default() -> Self {
        Self {
            inner: ::sys::default_levenberg_marquardt_params(),
        }
    }
}

impl LevenbergMarquardtParams {
    pub fn set_max_iterations(&mut self, value: u32) {
        ::sys::levenberg_marquardt_params_set_max_iterations(self.inner.pin_mut(), value)
    }
    pub fn set_lambda_upper_bound(&mut self, value: f64) {
        ::sys::levenberg_marquardt_params_set_lambda_upper_bound(self.inner.pin_mut(), value)
    }
    pub fn set_lambda_lower_bound(&mut self, value: f64) {
        ::sys::levenberg_marquardt_params_set_lambda_lower_bound(self.inner.pin_mut(), value)
    }
    pub fn set_diagonal_damping(&mut self, flag: bool) {
        ::sys::levenberg_marquardt_params_set_diagonal_damping(self.inner.pin_mut(), flag)
    }
    pub fn set_relative_error_to_l(&mut self, value: f64) {
        ::sys::levenberg_marquardt_params_set_relative_error_to_l(self.inner.pin_mut(), value)
    }
    pub fn set_absolute_error_to_l(&mut self, value: f64) {
        ::sys::levenberg_marquardt_params_set_absolute_error_to_l(self.inner.pin_mut(), value)
    }

}
